package/MDAnalysis/analysis/waterdynamics.py

```diff
@@ -1208,87 +1208,84 @@ class SurvivalProbability(object):
-            a = self._getOneDeltaPoint(selection1, totalFrames, frame, wint)
-            sumDeltaP += a
+        for frame_no in range(len(selected) - window_size):
+            delta = self._getOneDeltaPoint(selected, frame_no, window_size)
+            sumDeltaP += delta
             n += 1
 
-        # if no water molecules remain in selection, there is nothing to get
-        # the mean, so n = 0.
-        return sumDeltaP/n if n > 0 else 0
+        return sumDeltaP/n
 
-    def _NumPart_tau(self, selection, totalFrames, t0, tau):
+
+    def _getOneDeltaPoint(self, selected, t, tau):
         """
-        Compares the molecules in t0 selection and t0+tau selection and
-        select only the particles that remain from t0 to t0+tau. It returns
-        the number of remaining particles.
+        Gives one point to calculate the mean and
+        gets one point of the plot C_vect vs t.
+        - Ex: t0=1 and tau=1 so calculate
+        how many selected beads survive from the frame 1 to 2
+        - Ex: t0=5 and tau=3 so calculate
+        how many selected beads survive from the frame 5 to 8
         """
-        a = set(selection[t0])
-        i = 0
-        while (t0 + i) < t0 + tau and (t0 + i) < totalFrames:
-            b = set(selection[t0 + i])
-            a = a.intersection(b)
-            i += 1
-        return len(a)
-
-    def _NumPart(self, selection, t):
-        return len(selection[t])
 
-    def _selection_serial(self, universe, selection_str):
-        selection = []
-        pm = ProgressMeter(universe.trajectory.n_frames,
-                           interval=10, verbose=True)
-        for ts in universe.trajectory:
-            selection.append(universe.select_atoms(selection_str))
-            pm.echo(ts.frame)
-        return selection
+        Nt = len(selected[t])
+        if Nt == 0:
+            return 0
 
-    def run(self, **kwargs):
-        """Analyze trajectory and produce timeseries"""
+        Ntau = self._NumPart_tau(selected, t, tau)
+        return Ntau/Nt
 
-        # All the selection to an array, this way is faster than selecting
-        # later.
-        if self.nproc == 1:
-            selection_out = self._selection_serial(
-                self.universe, self.selection)
-        else:
-            # selection = selection_parallel(universe, selection_str, nproc)
-            # parallel selection to be implemented
-            selection_out = self._selection_serial(
-                self.universe, self.selection)
-        self.timeseries = []
-        for dt in list(range(1, self.dtmax + 1)):
-            output = self._getMeanOnePoint(
-                self.universe, selection_out, self.selection, dt, self.tf)
-            self.timeseries.append(output)
 
+    def _NumPart_tau(self, selected, t, tau):
+        """
+        Compares the molecules in t selection and t+tau selection and
+        select only the particles that remain from t to t+tau and
+        at each point in between.
```

Comment:
"and at each point in between" this is not necessary true, we just read the atoms in the "t" frame and in the "t+tau" frame, we don't know if atoms went out and re-enter in that elapsed time. Erase "and at each point in between".

Reply:
Thanks @alejob for the feedback. I initially thought it would be implemented this way (allowing re-entering). However, the function _NumPart_tau actually iterates over all frames in between t and t+tau and removes atoms which left the frame. That is done using the set.intersection() with each intermediate frame in the while loop. This is why I added this line to clarify it.

Reply:
You're right, I got confused with an other implementation of this module (intermittent HBL). Please just keep your comment.